### ISSUE TITLE
group trashes should exist?

### ISSUE THREAD

**Issue by mstfdkmn** [irods/irods#6901]
### iRODS Version, OS and Version
4.2.11 centos7, 4.3.0 almalinux8

### What did you try to do?
Without using force option I removed an object in the group collection that I am member of. 

### Expected behavior
I expect that object should be moved to the trash collection of the group.

### Observed behavior (including steps to reproduce, if applicable)
The object has not been moved to the trash collection of the group but to the trash collection of the user.

to reproduce:
```
iadmin mkgroup mygroup
iadmin atg mygroup rods
icd /u0137480/home/mygroup
iput test.txt
/u0137480/home/mygroup:
  test.txt
-bash-4.2$ irm test.txt
-bash-4.2$ ils /u0137480/trash/home/mygroup
/u0137480/trash/home/mygroup:
-bash-4.2$ ils /u0137480/trash/home/rods
/u0137480/trash/home/rods:
  C- /u0137480/trash/home/rods/mygroup
-bash-4.2$ ils /u0137480/trash/home/rods/mygroup
/u0137480/trash/home/rods/mygroup:
  test.txt
```
I am confused with this. Is this an expected behavior? If yes, then wat is the role of the `trash` collections of groups.

**Comment by alanking** [comment]
In my view, this is the expected behavior because no action is taken *as* a group in iRODS. Only users perform actions. In this case, `irm` is performed by `rods`, so the data object goes into the trash collection of `rods`. The reason the trash collection exists for groups is because of #2978. I think you can modify the `acCreateUserZoneCollections` policy to not create a trash collection for `rodsgroup`s.

Please correct me if I'm wrong, but the only reason we would expect the object to go into the group's trash collection in this instance is because it happened to exist in the group's home collection. If `rods` is a member of multiple groups (and I'm guessing `rods` is a member of group `public`, which is the default behavior), to which group's trash collection would a deleted data object go? What if the data object was in a collection not affiliated with either group?

You could implement policy to detect the collection and move to the desired trash collection, but I'm not sure if tools like `irmtrash` would then work as expected.

Hope that helps!

**Comment by alanking** [comment]
I guess to answer the issue title question directly: I personally don't think it makes sense for group trash collections to exist, but I also do not fully understand their role at this point in time. We could remove them by default, but that might constitute a "major" change. I defer to others on that point.

**Comment by trel** [comment]
I agree - while the current behavior isn't quite obvious - it's not clear how to make it much better either.   Not creating a home collection for rodsgroups by default is definitely a possibility... and doesn't close the door on an admin making any other collection later either...

Anyone else in favor / against not creating `rodsgroup` home collections by default?

**Comment by mstfdkmn** [comment]
> Please correct me if I'm wrong, but the only reason we would expect the object to go into the group's trash collection in this instance is because it happened to exist in the group's home collection. If `rods` is a member of multiple groups (and I'm guessing `rods` is a member of group `public`, which is the default behavior), to which group's trash collection would a deleted data object go? What if the data object was in a collection not affiliated with either group?
> 
Yes, I would expect the object to go into the group's trash collection if it is deleted from somewhere inside the group's home collection. In case a user is a member of multiple groups, I dont see any issue because depending on where the data object is stored,  once deleted I believe it should go to the trash of that group.

> What if the data object was in a collection not affiliated with either group?
I am not sure about this. Might it be a case that a data object exists in a collection not affiliated with either a group or a user?

Aside from what is being discussed my other concern is how to consider those data objects moved to trashes in terms of quota policies. If for a group a logical quota is applied on the home collection of that group, how are we going to monitor count deleted objects (objects moved to trashes)

**Comment by trel** [comment]
> Yes, I would expect the object to go into the group's trash collection if it is deleted from somewhere inside the group's home collection.

But, again, the group didn't delete the object - a user did.  So the data object gets 'moved' into that user's trash (on all resources where there is a replica).

> If for a group a logical quota is applied on the home collection of that group, how are we going to monitor/count objects deleted (objects moved to trashes)?

Since the logical quota is only looking at the collection, once something is in the trash, it will no longer be counted by that policy.

You could have a separate logical quota watching particular collections in the trash.

**Comment by mstfdkmn** [comment]
Alright, as Alan said above groups and users as entity are not same but they have similar features confusing.
As for the logical quota I was more thinking to apply it on a group collection and the trash collection of that group instead of trashes of each individual users, but okay. Thanks.

Then not creating rodsgroup home collections by default looks fine.

**Comment by alanking** [comment]
> I guess I dint get this completely Could it be possible that a data object exists in a collection not affiliated with either a group or a user?

By this I meant some collection which does not exist under `home`, or perhaps another user's collection which has granted permissions for other users. So, something like like `/tempZone/projects/coolscience` which is not even under the `home` collection would work here as an example. We could say that this collection is not affiliated with any group or user in particular, although users and groups would have permissions to work in said collection.

> Yes, I would expect the object to go into the group's trash collection if it is deleted from somewhere inside the group's home collection. In case a user is a member of multiple groups, I dont see any issue because depending on where the data object is stored, once deleted I believe it should go to the trash of that group.

I see. So, more generally, would it be fair to say that you would expect the `trash` collection to mirror the collection structure of the deleted contents? In other words, if we have `/tempZone/projects/coolscience/`, we would expect `/tempZone/trash/projects/coolscience` to appear when `/tempZone/projects/coolscience/` is `irm`'d, rather than the `trash/home` collection of the user performing the action?

Also... I feel like *home collections* for groups still make sense to create by default, but perhaps not *trash collections* (as we have established that nothing gets sent there... yet).

**Comment by korydraughn** [comment]
> Also... I feel like home collections for groups still make sense to create by default

Why do you feel that is acceptable? Have you identified a use-case for that?

**Comment by trel** [comment]
Creating them seems to send a mixed signal - I'm currently team no-group-home-collection-by-default.

**Comment by alanking** [comment]
I think I got confused about how we jumped from talking about trash collections to home collections, but that has cleared up.

I guess I don't have a use case in mind but was mainly thinking about how we use the `public` group home collection in our testing. We just need to make sure that we manually create that collection as part of our test setup logic whenever it is needed.

**Comment by korydraughn** [comment]
Agreed.

**Comment by mstfdkmn** [comment]
> By this I meant some collection which does not exist under `home`, or perhaps another user's collection which has granted permissions for other users. So, something like like `/tempZone/projects/coolscience` which is not even under the `home` collection would work here as an example. We could say that this collection is not affiliated with any group or user in particular, although users and groups would have permissions to work in said collection.

I see - what I was assuming for this scenario was that depending n the permission type (user based or group based) the deleted object could be counted for that entity (user or group). And;

- If a user is granted for a collection, a delete there goes to the user's trash.
- If a user is granted for a collection by his group permission, a delete there goes to the group's trash since the user deletes this for his group right.
- if a user is granted for a collection by both his group permission and his user permission, then ? and I was believing a delete should go to the user's trash since there is an individual/specific access right next to the group right.

But okay. On my end the conclusion (according to what discussed above) seems that;
- Since groups as an entity don't delete seeing deleted objects get 'moved' into users' trash is an expected behavior.
- As for the question of group trashes should exist, you think of changing the default policy (including both home and trash).

Thanks.

**Comment by korydraughn** [comment]
Based on the conversation, we'll adjust the server so that the creation of a group does not lead to the creation of a home collection for that group.

**Comment by alanking** [comment]
After some discussion, we have decided / re-affirmed the following:
1. When a user puts something in the trash (e.g. `irm`), the data should go into that user's trash collection. We can further discuss whether matching the original path in the trash path should be implemented, but we have found the current behavior as reported above to be acceptable.
2. Due to (1), group trash should not exist because groups cannot perform actions.
3. Groups having a home collection by default is fine/expected at this point. If the home collection is not desired, policy can be implemented to remove the collection immediately upon creation of the group.

These describe the current behavior of the system, so there is no behavioral change needed. However, we have decided that we should document the trash behavior in detail so it can be more readily understood.

@korydraughn / @trel  - Please amend as desired.

**Comment by trel** [comment]
Yes - seems good and correct.

The alternative course of keeping/forcing fullpaths to remain the 'same' on their way into the trash (and not putting them into the user's trash who performed the operation) would cause permissions issues.
